lib/puppetserver/ca/puppet_config.rb

```diff
@@ -0,0 +1,121 @@
+
+module Puppetserver
+  module Ca
+    # Provides an interface for asking for Puppet[ Server] settings w/o loading
+    # either Puppet or Puppet Server. Includes a simple ini parser that will
+    # ignore Puppet's more complicated conventions.
+    class PuppetConfig
+
+      def self.parse(config_path = nil)
+        instance = new(config_path)
+        instance.load
+
+        return instance
+      end
+
+      attr_reader :errors, :settings
+
+      def initialize(supplied_config_path = nil)
+        @using_default_location = !supplied_config_path
+        @config_path = supplied_config_path || user_specific_conf_file
+
+        @settings = nil
+        @errors = []
+      end
+
+      # Return the correct confdir. We check for being root (user id == 0)
+      # on *nix, else the user path. We do not include a check for running
+      # as Adminstrator since non-development scenarios for Puppet Server
+      # on Windows are unsupported.
+      def user_specific_conf_dir
+        if Gem.win_platform? && Process::UID.eid == 0
+          '/etc/puppetlabs/puppet'
+        else
+          "#{ENV['HOME']}/.puppetlabs/etc/puppet"
+        end
+      end
+
+      def user_specific_conf_file
+        user_specific_conf_dir + '/puppet.conf'
+      end
+
+      def load
+        unless @using_default_location && !File.exist?(@config_path)
+          results = parse_text(File.read(@config_path))
+        end
+
+        results ||= {}
+        results[:main] ||= {}
+        results[:master] ||= {}
+
+        overrides = results[:main].merge(results[:master])
+
+        @settings = resolve_settings(overrides).freeze
+      end
+
+      # Resolve the cacert, cakey, and cacrl settings from default values,
+      # with any overrides for the specific settings or their dependent
+      # settings (ssldir, cadir) taken into account.
+      def resolve_settings(overrides = {})
+        unresolved_setting = /\$[a-z_]+/
+
+        # Returning the key for unknown keys (rather than nil) is required to
+        # keep unknown settings in the string for later verification.
+        substitutions = Hash.new {|h, k| k }
+        settings = {}
+
+        confdir = user_specific_conf_dir
+        settings[:confdir] = substitutions['$confdir'] = confdir
+
+        ssldir = overrides.fetch(:ssldir, '$confdir/ssl')
+        settings[:ssldir] = substitutions['$ssldir'] = ssldir.sub('$confdir', confdir)
+
+        cadir = overrides.fetch(:cadir, '$ssldir/ca')
+        settings[:cadir] = substitutions['$cadir'] = cadir.sub(unresolved_setting, substitutions)
+
+        settings[:cacert] = overrides.fetch(:cacert, '$cadir/ca_crt.pem')
+        settings[:cakey] = overrides.fetch(:cakey, '$cadir/ca_key.pem')
+        settings[:cacrl] = overrides.fetch(:cacrl, '$cadir/ca_crl.pem')
+
+        settings.each_pair do |key, value|
+          settings[key] = value.sub(unresolved_setting, substitutions)
```

Comment:
I'm not great with regular expressions.  Could you tell me what's happening here/why? ```value.sub(unresolved_setting, substitutions)```

Reply:
Sorry, if this is overexplaining, this is what the block is doing, what sub is doing, and what the regex does.

The block is going over the so-far-computed values of `cacert`, `cakey`, and cacrl and interpolating any references to `$cadir`, `$ssldir`, or `$confdir` within them to their computed values.

`sub` will take a regex, if it has a match it will look up the match in a hash if that's passed in as a second arguement, replacing the matched substring with the value at that key in the hash.

eg.
```
substitutions = {"$confdir" => '/var/lib'}
regex = /<regex-that-matches-$confdir>/
"$confdir/foo/bar".sub(regex, substitutions)
  => "/var/lib/foo/bar"
```

And the regex itself should match a valid puppet setting that starts with a '$' (including the dollar sign) and ends with '/' (excluding the slash). The regex,
`/\$[a-z_]+/`, will match a dollar sign followed by one or more of lowercase letters or an underscore.